File: Linked_list.cpp


#include <bits/stdc++.h>
using namespace std;

/* Link list node */
class Node {
public:
	int data;
	Node* next;
};


void rotate(Node** head_ref, int k)
{
	if (k == 0)
		return;

	
	// list = 10->20->30->40->50->60.
	Node* current = *head_ref;

	int count = 1;
	while (count < k && current != NULL) {
		current = current->next;
		count++;
	}


	if (current == NULL)
		return;

	Node* kthNode = current;

	while (current->next != NULL)
		current = current->next;


	current->next = *head_ref;

	*head_ref = kthNode->next;


	kthNode->next = NULL;
}


void push(Node** head_ref, int new_data)
{

	Node* new_node = new Node();

	new_node->data = new_data;


	new_node->next = (*head_ref);

	
	(*head_ref) = new_node;
}


void printList(Node* node)
{
	while (node != NULL) {
		cout << node->data << " ";
		node = node->next;
	}
}


int main(void)
{

	Node* head = NULL;

	// create a list 10->20->30->40->50->60
	for (int i = 60; i > 0; i -= 10)
		push(&head, i);

	cout << "Given linked list \n";
	printList(head);
	rotate(&head, 4);

	cout << "\nRotated Linked list \n";
	printList(head);

	return (0);
}

File: LinkedList.java
public class LinkedList{
    
    Node head;
	Node current;
public static class Node{
		int data;
		Node next;
Node(int d){
			data =d; next=null;
				}
	}
public void insertAtFirst(int d){
		if(head ==null)
		{
			head = new Node(d);
			
		}
		else{
			Node n = new Node(d);
			n.next =head;
			head =n;
		}
	
	}

	public void insertAfter(int n1,int d){
		Node temp =head;
		while(temp!=null){
			if(temp.data==n1){
				Node n =new Node(d);				
				n.next =temp.next;
				temp.next =n;
				break;

			}else{
				temp=temp.next;
			}
		}
	}

	public void insert(int d){
		if(head==null)
		{
			System.out.println("out");
			head =new Node(d);
			current =head;
		}else{
			
			System.out.println("in ");
			Node n = new Node(d);
			current.next=n;
			current=current.next;
        }

	}

	public void delete(int d){
	Node currentNode=head;
		Node previousNode=null;
		if(head.data ==d){
			head = head.next;
		}else{
			while(currentNode!=null){
				if(currentNode.data==d){
					previousNode.next=currentNode.next;
					break;
				}else{
				previousNode=currentNode;
				currentNode =currentNode.next;
			}
			}
		}

	}
	 public void show(){
	 	System.out.println("---Showing---");
	 	Node myNode =head;
	 	while(myNode!=null){
	 		System.out.println(myNode.data);
	 		myNode =myNode.next;
	 	}
	 }	

	 public void deleteAtPosition(int d){

	 	Node temp =head;
	 	Node prev =null;
	 	int count=1;
		if(d==1){
	 			head =head.next;

	 		}else{
	 	while(temp!=null)
	 	{
	 		if(count==d){
	 			prev.next=temp.next;

	 			break;
	 		}else{
	 			prev =temp;
	 			temp =temp.next;
	 			count++;

	 		}
	  	}
	 }
	 }
}